src/test/java/org/testng/internal/DynamicGraphHelperTest.java

```diff
@@ -130,6 +134,16 @@ public void testCreateDynamicGraphWithGroupByInstances() {
     assertThat(actualMethodNames).contains("testMethod", "anotherTestMethod");
   }
 
+  @Test
+  public void testCreateDynamicGraphWithPackageWithAbstractClass() {
+    Class<?>[] classes =
+            new Class<?>[] {PackageTestClassA.class, PackageTestClassBAbstract.class, PackageTestClassBB.class};
+    XmlTest xmlTest = createXmlTest("2249_suite", "2249_test", classes);
+    DynamicGraph<ITestNGMethod> graph = newGraph(xmlTest, classes);
+    assertThat(graph.getFreeNodes().stream().map(ITestNGMethod::getMethodName).collect(Collectors.toList()))
+            .containsExactly("a1", "a2");
```

Comment:
Any idea why b1 and b2 are not in the list?
What was the result before the fix?

Reply:
@sergejjkim - I also have the same question as @juherr . As per your suite file, only `nevermind()` should have gotten excluded. Why is `b1` and `b2` also getting excluded from `freeNodes()` when they should have been eligible for execution too ?

Reply:
I think I'm missing something because I don't understand why dependency is expected here.
There is no rule on the execution order. So, any method of the 4 is legit to be run.

In fact, with or without PackageTestClassBAbstract in the run, you must have the same result.
Could you add the test case without PackageTestClassBAbstract?

Reply:
> I think I'm missing something because I don't understand why dependency is expected here.
There is no rule on the execution order. So, any method of the 4 is legit to be run.

@juherr with or without this pull request DynamicGraphHelper builds dependencies in createDynamicGraph() method, specifically by calling 
`ListMultiMap<ITestNGMethod, ITestNGMethod> classDependencies =
          createClassDependencies(methods, xmlTest);`

> In fact, with or without PackageTestClassBAbstract in the run, you must have the same result.

Exactly, this is the whole point of this pull request. Before it you would get 2 free nodes without PackageTestClassBAbstract (the number of test methods in 1st test class) and 4 with PackageTestClassBAbstract (sum of numbers of tests methods in 1st test class and the one after the abstract one). Now you are supposed to get 2 in both cases.

> Could you add the test case without PackageTestClassBAbstract?
Sure!

Reply:
so i've added test without abstract class. both of them expect to get 2 methods: a1 and a2 as getFreeNodes(), please see results below:
![image](https://user-images.githubusercontent.com/11964660/75477163-8b039d80-59a4-11ea-941c-d5cee38c03c3.png)

Reply:
Ok, I got it!
b1 & b2 are only expected due to preserve-order which is true by default
https://github.com/cbeust/testng/blob/2e8a95bca5b92a063a3abdf9e28b4e9fddb482ac/src/main/java/org/testng/xml/XmlSuite.java#L203

Can I ask you to confirm that both tests are returning the 4 expected methods with preserve-order to false?

Reply:
@krmahadevan I think we can merge

Reply:
@juherr 

> Can I ask you to confirm that both tests are returning the 4 expected methods with preserve-order to false?

sure, i've reworked unit tests to confirm that, so there are tests with and without abstract class returning 2 if preserve-order == true and 4 otherwise

Reply:
great 👍